### PR TITLE
[Disk Manager] should convert operation error to disk manager api format

### PR DESCRIPTION
**Pull request by BarkovBG**

### CODE REVIEW

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 1]
```diff
@@ -3,10 +3,13 @@ package facade
 import (
 	"context"
 
-	"github.com/ydb-platform/nbs/cloud/disk_manager/api"
+	disk_manager "github.com/ydb-platform/nbs/cloud/disk_manager/api"
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:6]: ?

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 2]
```diff
@@ -34,8 +37,36 @@ func getOperation(
 
 	switch result := o.Result.(type) {
 	case *operation.Operation_Error:
+		status := grpc_status.FromProto(result.Error)
+		dmStatus := grpc_status.New(status.Code(), status.Message())
+
+		// We should return disk_manager.ErrorDetails, comply with contract with
+		// compute.
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:44]: // Convert to disk_manager.ErrorDetails via marshaling/unmarshaling in order to be compliant with disk_manager/api

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 3]
```diff
@@ -34,8 +37,36 @@ func getOperation(
...
+		for _, errorDetail := range status.Details() {
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:45]: errorDetails

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 4]
```diff
@@ -34,8 +37,36 @@ func getOperation(
...
+			protoMessage, err := proto.Marshal(errorDetail.(proto.Message))
+			if err != nil {
+				logging.Warn(ctx, "failed to marshal error detail: %v", err)
+				continue
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:49]: Lets expect that we have single error details here

File: cloud/disk_manager/internal/pkg/facade/testcommon/common.go [hunk 5]
```diff
@@ -11,11 +11,10 @@ import (
 	"testing"
 	"time"
 
-	"github.com/golang/protobuf/proto"
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/google/uuid"
 	"github.com/stretchr/testify/require"
-	"github.com/ydb-platform/nbs/cloud/disk_manager/api"
+	disk_manager "github.com/ydb-platform/nbs/cloud/disk_manager/api"
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/testcommon/common.go:17]: ?

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 6]
```diff
@@ -34,8 +37,36 @@ func getOperation(
...
+			}
+
+			dmStatusWithDetails, err := dmStatus.WithDetails(&dmErrorDetail)
+			if err == nil {
+				dmStatus = dmStatusWithDetails
+			} else {
+				logging.Warn(ctx, "failed to attach error details: %v", err)
+				continue
+			}
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:65]: Better to move this code to separate func

File: cloud/disk_manager/internal/pkg/facade/operation_service.go (hunk 1 above)
Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/facade/operation_service.go:6]: https://github.com/ydb-platform/nbs/pull/1153

File: cloud/disk_manager/internal/pkg/facade/testcommon/common.go (hunk 5 above)
Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/facade/testcommon/common.go:17]: https://github.com/ydb-platform/nbs/pull/1153

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 7]
```diff
@@ -5,12 +5,60 @@ import (
...
+	"github.com/ydb-platform/nbs/cloud/tasks/logging"
 	"github.com/ydb-platform/nbs/cloud/tasks/operation"
 	"google.golang.org/grpc"
+	grpc_status "google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
 )
 
 ////////////////////////////////////////////////////////////////////////////////
 
+func convertOperationErrorToDiskManagerApiFormat(
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:17]: toDiskManagerOperationError

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 8]
```diff
@@ -5,12 +5,60 @@ import (
...
+
+	// Convert to disk_manager.ErrorDetails via marshaling/unmarshaling in
+	// order to be compliant with disk_manager/api.
+	protoMessage, err := proto.Marshal(errorDetail.(proto.Message))
+	if err != nil {
+		logging.Warn(ctx, "failed to marshal error detail: %v", err)
+		return defaultDmOperationError
+	}
+
+	var dmErrorDetail disk_manager.ErrorDetails
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:44]: dmErrorDetails

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 9]
```diff
@@ -5,12 +5,60 @@ import (
...
+	err = proto.Unmarshal(protoMessage, &dmErrorDetail)
+	if err != nil {
+		logging.Warn(ctx, "failed to unmarshal error detail: %v", err)
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:47]: error details

File: cloud/disk_manager/internal/pkg/facade/operation_service.go [hunk 10]
```diff
@@ -5,12 +5,60 @@ import (
...
+	}
+
+	// We expect the only one error detail.
+	errorDetail := status.Details()[0]
+
+	// Convert to disk_manager.ErrorDetails via marshaling/unmarshaling in
+	// order to be compliant with disk_manager/api.
+	protoMessage, err := proto.Marshal(errorDetail.(proto.Message))
+	if err != nil {
+		logging.Warn(ctx, "failed to marshal error detail: %v", err)
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/facade/operation_service.go:40]: error details

**Review by SvartMetal — APPROVED** [review]

**Review by leftmain — APPROVED** [review]

**Review by gy2411 — APPROVED** [review]

**Pull request merged**